src/credentials/providers/ec2CredentialsProvider.ts

```diff
@@ -1,27 +1,56 @@
 /*!
- * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
  * SPDX-License-Identifier: Apache-2.0
  */
 
-import * as AWS from 'aws-sdk'
-import { CredentialType } from '../../shared/telemetry/telemetry.gen'
+import { Credentials, EC2MetadataCredentials, MetadataService } from 'aws-sdk'
 import { getStringHash } from '../../shared/utilities/textUtilities'
-import { CredentialsProvider, CredentialsProviderType ,CredentialsId } from './credentials'
-
+import { getLogger } from '../../shared/logger'
+import { CredentialType } from '../../shared/telemetry/telemetry.gen'
+import { CredentialsId, CredentialsProvider, CredentialsProviderType } from './credentials'
 
 /**
  * Credentials received from EC2 metadata service.
  *
  * @see CredentialsProviderType
  */
 export class Ec2CredentialsProvider implements CredentialsProvider {
-    public constructor(private token: string) {
+    private static readonly METADATA_SERVICE_TIMEOUT: number = 500
+
+    private credentials: EC2MetadataCredentials | undefined
+    private region: string | undefined
+    private available: boolean | undefined
+
+    public constructor(private metadata: MetadataService = Ec2CredentialsProvider.makeMetadataService()) {}
+
+    public isAvailable(): Promise<boolean> {
+        // this check is only performed once per activation
+        if (this.available !== undefined) {
+            return Promise.resolve(this.available)
+        }
+
+        return new Promise(resolve => {
+            const start = Date.now()
+            this.metadata.request('/latest/dynamic/instance-identity/document', (err, response) => {
+                if (response) {
+                    const document = JSON.parse(response)
+                    this.region = document['region']
+                    getLogger().verbose(`resolved instance region ${this.region} from EC2 metadata`)
+                }
+
+                const elapsed = Date.now() - start
+                getLogger().verbose(`took ${elapsed}ms to call the EC2 metadata service`)
```

Comment:
see [guidelines](https://github.com/aws/aws-toolkit-vscode/blob/84a6d4ae45fd545ae04a7805e0d918b55d463860/docs/CODE_GUIDELINES.md) for suggested log message structure. It's useful to put the "variable" part at the end of the message. And to have some consistent structure, if possible (such as a common topic prefix).

```suggestion
                getLogger().verbose(`credentials: EC2 metadata service call took ${elapsed}ms`)
```